openssl_pkcs12: Add idempotency checks

Shaps opened the pull request:
##### SUMMARY
Adds idempotency checks to the openssl_pkcs12 module and related tests. 
Also decoupled `parse` and `generate` from the file write, as they are now used for multiple stuff that don't require the file to be written to disk.
Fixes #53221

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
openssl_pkcs12

Commit:
Added idempotency logic to openssl_pkcs12

Also decoupled the 'parse' and 'generate' function from the file write
as they are now used in different places that do not need the file to be
written to disk.

Commit:
Added idempotency tests for openssl_pkcs12

Also adds a new test for pkcs12 files with multiple certificates

felixfontein commented on lib/ansible/modules/crypto/openssl_pkcs12.py:
```diff
@@ -222,7 +222,51 @@ def _check_pkey_passphrase():
+        if os.path.exists(self.path) and module.params['action'] == 'export':
+            dummy = self.generate(module)
+            self.src = self.path
+            pkcs12_privatekey, pkcs12_certificate, pkcs12_other_certificates, pkcs12_friendly_name = self.parse()
```
You should wrap this in a `try/except` and return `False` if it raises an exception.

Commit:
Regenerate if parsed file is invalid

Commit:
pkcs12_other_certificates check was wrong

felixfontein commented on lib/ansible/modules/crypto/openssl_pkcs12.py:
```diff
@@ -237,7 +237,54 @@ def _check_pkey_passphrase():
+                                                        self.pkcs12.get_certificate())
+                if pkcs12_certificate != expected_cert:
+                    for pkcs_cert in pkcs12_other_certificates:
+                        if pkcs_cert == expected_cert:
```
Hmm, if the certificate is found in `pkcs12_other_certificates`, I think you still have to return `False`, since you want the certificate to be the leaf certificate and not part of the chain.

Shaps replied:
In my tests, the certs are not necessarily returned in the correct order, sometimes the cert ends up in the `pkcs12_other_certificates` reason why this check. To be fair it does make sense in a way, in the end a pkcs12 file will just be a list of certs bundled together, it's pyopenssl that calls them 'ca_certificates' but they don't need to necessarily be CA certs, they can be any cert.

felixfontein replied:
The distinction between cert and CA cert is made by OpenSSL itself: https://www.openssl.org/docs/manmaster/man3/PKCS12_create.html https://www.openssl.org/docs/manmaster/man3/PKCS12_parse.html

That's really strange, since OpenSSL both writes and reads the PKCS12 file, I would expect the order to be preserved.

felixfontein replied:
I've been playing around with this a bit; the problem seems to be loading the PKCS12 file with OpenSSL. When simply dumping the result of both runs via `openssl pkcs12 -info`, the output is the same for both. Also, the certificates show up in the correct order, and friendly names are correct. I'll go look at how OpenSSL loads PKCS12 files next...

felixfontein replied:
Ok, I think I got it :-) The problem is that in your test, you use the same private key for **all** certificates. This won't happen in real life if you have a bunch of different certificates. The OpenSSL code is here: https://github.com/openssl/openssl/blob/master/crypto/pkcs12/p12_kiss.c

It works as follows:
 - crawl through PKCS12 file and collect private key and all certs (in correct order)
 - pop certs from cert stack (i.e. reverse order!), and:
   - use first cert that matches private key as main cert;
   - if there is no private key, or cert doesn't match private key, or already a cert matching the private key was found, add the cert to CA certs stack.

This results in the following:
 - all the problems you described are caused by all three certs in the .p12 file using the same private key;
 - in general, if no private key is in the .p12 file, friendly name isn't parsed correctly (because the main certificate itself won't be found);
 - the CA certs are returned in reverse order as specified on creation.

This also explains why there was this funny condition that friendly name and private key path were required together (which we removed in #54370): that's because OpenSSL totally screws up otherwise! (I guess this is why that restriction was introduced in the first place. That should have *really* been documented...)

We can obviously work around this, at least partially, but *only* assuming that the handling of pyOpenSSL and OpenSSL of PKCS12 files will never change. This is really !#(*$#@!(*!!@. I guess the main result is "don't use OpenSSL for PKCS12 file handling if possible". Anyway, assuming that pyOpenSSL won't change is pretty safe, since it's mostly deprecated anyway (except for handling TLS connections), this part of the code probably won't change. Well, probably. Who knows. Then there is pyca/pyopenssl#770, in which it is said `"We would accept a PR here to expand support [...]"`.

felixfontein replied:
About the naming `ca_certificates` vs. `other_certificates`: how about renaming the option to `other_certificates` (and keeping `ca_certificates` as an alias)? That should reduce confusion a lot.

Shaps replied:
This should hopefully be fixed now

felixfontein commented on lib/ansible/modules/crypto/openssl_pkcs12.py:
```diff
@@ -237,7 +237,54 @@ def _check_pkey_passphrase():
+            if (pkcs12_other_certificates is not None) and (self.ca_certificates is not None):
+                expected_ca_certs = [crypto.dump_certificate(crypto.FILETYPE_PEM,
+                                                             ca_cert) for ca_cert in self.pkcs12.get_ca_certificates()]
+                expected_ca_certs.append(crypto.dump_certificate(crypto.FILETYPE_PEM, self.pkcs12.get_certificate()))
```
Why do you include the current certificate in the set of CA certs?

Shaps replied:
Same reason as above

felixfontein commented on lib/ansible/modules/crypto/openssl_pkcs12.py:
```diff
@@ -237,7 +237,54 @@ def _check_pkey_passphrase():
+            elif bool(pkcs12_other_certificates) != bool(self.ca_certificates):
+                return False
+
+            if not pkcs12_other_certificates:
```
What is this `if` for?

Shaps replied:
It appears that if the pkcs12 file has additional certs, the `get_friendlyname` method ( from `parse` ) returns None, making this fail unnecessarily. That said, I guess I could/should change this to be 
```python
not pkcs12_other_certificates and not pkcs12_friendly_name
```

felixfontein replied:
Hmm, that's even more strange. I'd say outright terrible :)

I've looked at the pyOpenSSL implementation of encoding and decoding PKCS12 files, it is essentially calling the OpenSSL functions from the other comment. The friendly name comes from the main cert extracted from the PKCS12 file, i.e. when the order is f**ked up, I would expect `friendly_name` to also be wrong.

felixfontein commented on lib/ansible/modules/crypto/openssl_pkcs12.py:
```diff
@@ -237,7 +237,54 @@ def _check_pkey_passphrase():
+            dummy = self.generate(module)
+            self.src = self.path
+            try:
+                pkcs12_privatekey, pkcs12_certificate, pkcs12_other_certificates, pkcs12_friendly_name = self.parse()
```
Why not rename `pkcs12_other_certificates` to `pkcs12_ca_certs`?

Shaps replied:
because as above, they don't have to be ca certs, they may be ( and probably usually are ), but pkcs12 is fine with any bundle of certs.

felixfontein commented:
Also, you need a changelog fragment :)

Shaps commented:
YES :D
> Also, you need a changelog fragment :)

felixfontein commented on lib/ansible/modules/crypto/openssl_pkcs12.py:
```diff
@@ -237,7 +237,54 @@ def _check_pkey_passphrase():
+
+            if not pkcs12_other_certificates:
+                if ((self.pkcs12.get_friendlyname() is not None) and (pkcs12_friendly_name is not None)):
+                    return self.pkcs12.get_friendlyname() == pkcs12_friendly_name
```
BTW, currently the `_check_pkey_passphrase()` will only be called if friendly name is not set for both the PKCS12 file and in the module options. I guess this should better be
```suggestion
                    if self.pkcs12.get_friendlyname() != pkcs12_friendly_name:
                        return False
```

felixfontein commented on lib/ansible/modules/crypto/openssl_pkcs12.py:
```diff
@@ -363,10 +415,13 @@ def main():
                 else:
-                    pkcs12.parse(module)
+                    pkey, cert, ca_certs, friendly_name = pkcs12.parse()
+                    dump_content = '%s%s%s' % (to_native(pkey), to_native(cert), b''.join(ca_certs))
```
```suggestion
                    dump_content = '%s%s%s' % (to_native(pkey), to_native(cert), to_native(b''.join(ca_certs)))
```
Otherwise the last line of the PEM file will be `b'-----BEGIN CERTIFICATE-----\nMIIE4zCCAsugAwIBAgIUerXv6Q7bEPOTRmAehOZkVavpUz0wDQYJKoZIhvcNAQEL\nBQAwGzEZMBcGA1UEAwwQd3d3...` (at least in Python 3).

Commit:
Updated ca_certificates to other_certificates

ca_certificates is left as an alias to other_certificates;
friendlyname depends on private key, so it will be ignored while
checking for idempotency if the pkey is not set;
idempotency check only checks for correct certs in the stack

Commit:
use different keys for different certs

Commit:
Added other_certificates in module docs

Commit:
Added changelog and porting guide

felixfontein commented on lib/ansible/modules/crypto/openssl_pkcs12.py:
```diff
@@ -29,10 +29,11 @@
+    other_certificates:
         description:
-            - List of CA certificate to include.
+            - List of other certificates to include.
```
Maybe add a comment that this was called `ca_certificates` before Ansible 2.8

felixfontein commented on docs/docsite/rst/porting_guides/porting_guide_2.8.rst:
```diff
@@ -333,6 +333,12 @@ Noteworthy module changes
 * The ``win_dsc`` module will now validate the input options for a DSC resource. In previous versions invalid options
   would be ignored but are now not.
 
+* The ``win_domain_membership`` module will no longer automatically join a host in a domain that already has an account
```
This shouldn't be in this PR.

felixfontein commented on lib/ansible/modules/crypto/openssl_pkcs12.py:
```diff
@@ -237,7 +238,48 @@ def _check_pkey_passphrase():
+            elif bool(pkcs12_other_certificates) != bool(self.ca_certificates):
+                return False
+
+            if pkcs12_privatekey:
```
Maybe add a comment that OpenSSL only loads the friendly name if a private key is in the PKCS12 file.

felixfontein commented on lib/ansible/modules/crypto/openssl_pkcs12.py:
```diff
@@ -303,17 +338,29 @@ def parse(self, module):
                                           p12.get_privatekey())
             crt = crypto.dump_certificate(crypto.FILETYPE_PEM,
                                           p12.get_certificate())
+            ca_certs = []
```
I guess this should also be renamed?

Commit:
removed unrelated porting guide entry

Commit:
renamed ca_cert* occurrence with other_cert

felixfontein reviewed (approved):
shipit

(sorry for using wrong account...)

gundalow commented:
Thanks all!

felixfontein commented:
@Shaps thanks a lot for implementing this!!!
@gundalow thanks for merging!
